design/proposals/spike-cli-dockerless.md

```diff
@@ -0,0 +1,257 @@
+| Python CLI | Tested very well                                             |
+
+None of our CLIs test against other platforms besides Linux.
+
+|            | Security |
+| ---------- | -------- |
+| Ruby CLI   | See scan |
+| Go CLI     | See scan |
+| Python CLI | See scan |
+
+#### CLI commands to be supported
+
+The CLI should support the following commands:
+
+*Core* -  `init`, `authn` (`login`, `authenticate`),  `policy (load)`, `list`, `show `, `rotate_api_key` (`host` / `user`) and `fetch secrets`
+
+*Additional we might want to consider* - `resource`, `role`, `check`, `help`
+
+#### Estimation
+
+The following table summarizes the findings criteria detailed above and translates them to numbered equivalents. 
+
+1 - the least work
+
+3 - the most work
+
+|            | Functionality | Quality | UX   | Deployment | Testing | Security |
+| ---------- | ------------- | ------- | ---- | ---------- | ------- | -------- |
+| Ruby CLI   | 1             | 1       | 2    | 3          | 1       | See scan |
+| Go CLI     | 3             | 3       | 3    | 2          | 3       | See scan |
+| Python CLI | 2             | 2       | 1    | 1          | 2       | See scan |
+
+Ruby CLI ~11 weeks
+
+- Ramp-up on creating packages and create design for CLI (~1 week)
+- Provide package (rpm, pkg, deb, etc) / [deliver as executable](https://www.ruby-toolbox.com/categories/packaging_to_executables) (CLI + dependencies) (~3 week)
+- Expand current testing if needed (~1 week)
+- Build infrastructure for testing on new platforms (Windows, macOS, REHL environments) (~3 weeks)
+- Run a few CLI commands on new supported platforms (Windows, macOS environments) (~2 weeks)
+- Close open issues (~1 week)
+
+Python CLI ~13 weeks
+
+- Python ramp-up and create design for CLI (~1 week)
+- Implement logic to fill in functionality gaps (7 CLI commands, ~3 weeks)
+  - `init`, `authn` (`login`, `authenticate`), `show `, `rotate_api_key` (`host` / `user`)
+- Write associated tests for new implementation (~3 weeks)
+- Build infrastructure for testing on new platforms (Windows, macOS environments) (~3 weeks)
+- Run a few of them on different supported platforms (Windows, macOS, REHL environments) (~2 weeks)
+- Close open issues (~1 week)
+- Add documentation
+
+Go CLI - Complete rewrite (~18+ weeks)
+
+- Create design for CLI
+- CLI Ramp up
+- Redesign CLI
+- Implement new CLI commands and logic
+- Write tests for new functionality
+- Might need to redesign and build new infrastructure for environment testing (Windows, macOS, REHL environments)
+- Run a few CLI commands on new supported platforms (Windows, macOS, REHL environments)
+- Add documentation
+
+#### Takeways and recommendation
+
+Ruby requires the least amount of work in terms of what is already fuctionally supported but its UX is not intuitive. As part of its deployment, the CLI requires a specific version of Ruby installed on the machine to run the CLI.
```

Comment:
Although the interface is sub-optimal, why not launch with it (reduces time), and either refactor the interface in the future or focus on the GoLang CLI, which greatly simplify the CLI experience across all platforms.

Reply:
I am not so sure the amount of time that will be saved will be substantial when compared to the Python CLI and we could get a better UX. IMO changing the UX dramatically down the road might be confusing for the user and demand from us a lot more work. If we start with the Ruby CLI we will need to understand how to package it (single binary, zip, etc) which is added effort because from what I have seen isn't so straightforward.

Reply:
Is the outcome a CLI for the Dockerless DAP, or a new CLI?  It feels like there is some big batching going on by reworking the CLI for Dockerless.  Would it make more sense to push the CLI to a future phase?

Reply:
we need to provide our customers with a total dockerless deployement so that means the CLI. Users can use the current Ruby CLI and just install it via gem but that isn't good enough and not a solution that can work across the board. I did a minispike to see if we can somehow pack the ruby cli and get a single executable (see [here](https://github.com/cyberark/conjur/pull/1917/files#diff-94fda9f3ac9c67857646893fe6d3fd0053a6b319cd511dc03744f648e3da293dR163). For the reasons listed in the link we should not proceed with attempting to do so.